File: rust/core-bridge/tests/activity_protocol.rs
use ocaml_temporal_core_bridge::activity_protocol::{
    ActivityCancel, ActivityCancelReason, ActivityCompletion, ActivityCompletionResult,
    ActivityHeartbeat, ActivityTask, ActivityTaskVariant, completion_to_core, decode_completion,
    decode_heartbeat, decode_task, encode_completion, encode_heartbeat, encode_task,
    task_from_core,
};
use ocaml_temporal_core_bridge::workflow_protocol::Payload;
use std::collections::BTreeMap;
use temporalio_protos::coresdk::activity_task as core_activity_task;

/// A complete start-task fixture exercises every nullable field, the retry
/// policy, priority, headers, and both payload collection shapes.
const VALID_START_TASK: &str = r#"{"task_token":"AAEC/v8=","variant":{"kind":"start","workflow_namespace":"default","workflow_type":"example.workflow","workflow_execution":{"workflow_id":"workflow-1","run_id":"run-1"},"activity_id":"activity-1","activity_type":"example.activity","header_fields":{"trace":{"metadata":{},"data":{"encoding":"base64","data":"aGVhZGVy"}}},"input":[{"metadata":{"encoding":{"encoding":"base64","data":"anNvbi9wbGFpbg=="}},"data":{"encoding":"base64","data":"eyJ2YWx1ZSI6MX0="}}],"heartbeat_details":[],"scheduled_time":{"seconds":10,"nanoseconds":20},"current_attempt_scheduled_time":null,"started_time":{"seconds":11,"nanoseconds":0},"attempt":1,"schedule_to_close_timeout":{"seconds":60,"nanoseconds":0},"start_to_close_timeout":{"seconds":30,"nanoseconds":0},"heartbeat_timeout":null,"retry_policy":{"initial_interval":{"seconds":1,"nanoseconds":0},"backoff_coefficient_bits":"4611686018427387904","maximum_interval":{"seconds":60,"nanoseconds":0},"maximum_attempts":3,"non_retryable_error_types":["InvalidInput"]},"priority":{"priority_key":2,"fairness_key":"tenant","fairness_weight_bits":1065353216},"standalone_run_id":""}}"#;

/// Replaces exactly one fixture fragment so each malformed document targets a
/// single semantic rule and fails loudly if the fixture is later reshaped.
fn replace_once(source: &str, before: &str, after: &str) -> String {
    let offset = source
        .find(before)
        .unwrap_or_else(|| panic!("missing test fragment {before}"));
    let mut output = String::with_capacity(source.len() + after.len() - before.len());
    output.push_str(&source[..offset]);
    output.push_str(after);
    output.push_str(&source[offset + before.len()..]);
    output
}

/// Activity task tokens survive canonical JSON and return to Core unchanged.
#[test]
fn activity_completion_preserves_opaque_task_token() {
    let completion = ActivityCompletion {
        task_token: "AAEC/v8=".to_owned(),
        result: ActivityCompletionResult::Completed { result: None },
    };

    let encoded = encode_completion(&completion).expect("completion should encode");
    assert_eq!(decode_completion(&encoded), Ok(completion.clone()));
    assert_eq!(
        completion_to_core(&completion)
            .expect("completion should convert")
            .task_token,
        vec![0, 1, 2, 254, 255]
    );
}

/// The semantic activity decoder rejects unknown fields and non-canonical or
/// empty tokens before any completion can reach Core.
#[test]
fn activity_documents_are_closed_and_tokens_are_canonical() {
    assert!(
        decode_completion(
            r#"{"task_token":"AA==","result":{"kind":"will_complete_async"},"extra":true}"#
        )
        .is_err()
    );
    assert!(
        decode_completion(r#"{"task_token":"","result":{"kind":"will_complete_async"}}"#).is_err()
    );
    assert!(
        decode_completion(r#"{"task_token":"AA","result":{"kind":"will_complete_async"}}"#)
            .is_err()
    );
}

/// The tagged completion-result enum is itself closed, not just the wrapping
/// `ActivityCompletion` struct: an unknown member nested inside a `result`
/// variant's fields must fail closed the same way it does for every
/// analogous workflow-side tagged enum, instead of being silently dropped by
/// serde's default behavior.
#[test]
fn activity_completion_result_variants_reject_unknown_fields() {
    assert!(
        decode_completion(
            r#"{"task_token":"AA==","result":{"kind":"completed","result":null,"injected":true}}"#
        )
        .is_err()
    );
}

/// Heartbeats preserve the opaque token and binary detail payloads while
/// keeping the JSON object closed for forward-compatible validation.
#[test]
fn activity_heartbeat_round_trips_binary_details() {
    let mut metadata = BTreeMap::new();
    metadata.insert("encoding".to_owned(), b"binary/plain".to_vec());
    let heartbeat = ActivityHeartbeat {
        task_token: "AAEC/v8=".to_owned(),
        details: vec![Payload {
            metadata,
            data: vec![0, 1, 2, 254, 255],
        }],
    };

    let encoded = encode_heartbeat(&heartbeat).expect("heartbeat should encode");
    assert_eq!(decode_heartbeat(&encoded), Ok(heartbeat));
    assert!(encoded.contains("\"task_token\":\"AAEC/v8=\""));
    assert!(encoded.contains("\"details\""));
}

/// Proves that heartbeat details are copied into owned Rust values instead of
/// borrowing the temporary JSON input. The bridge must be able to retain the
/// decoded payload until Core consumes it after the parser's input buffer has
/// been released.
#[test]
fn activity_heartbeat_owns_input_after_decode() {
    let input = String::from(
        r#"{"task_token":"AAEC/v8=","details":[{"metadata":{"encoding":{"encoding":"base64","data":"YmluYXJ5L3BsYWlu"}},"data":{"encoding":"base64","data":"AAEC/v8="}}]}"#,
    );

    let heartbeat = decode_heartbeat(&input).expect("heartbeat should decode");
    drop(input);

    assert_eq!(heartbeat.task_token, "AAEC/v8=");
    assert_eq!(heartbeat.details.len(), 1);
    assert_eq!(heartbeat.details[0].data, vec![0, 1, 2, 254, 255]);
    assert_eq!(
        heartbeat.details[0]
            .metadata
            .get("encoding")
            .expect("encoding metadata should survive input drop"),
        &b"binary/plain".to_vec()
    );
}

/// The heartbeat decoder rejects unknown fields, empty/non-canonical tokens,
/// and payload encodings that are not the canonical binary representation.
#[test]
fn activity_heartbeat_rejects_unsafe_documents() {
    assert!(decode_heartbeat(r#"{"task_token":"AAEC/v8=","details":[],"extra":true}"#).is_err());
    assert!(decode_heartbeat(r#"{"task_token":"","details":[]}"#).is_err());
    assert!(decode_heartbeat(r#"{"task_token":"AAE","details":[]}"#).is_err());
    assert!(decode_heartbeat(
        r#"{"task_token":"AAEC/v8=","details":[{"metadata":{},"data":{"encoding":"raw","data":"AA=="}}]}"#
    )
    .is_err());
}

/// Cancellation tasks retain the stable reason while preserving the single
/// original task-token completion obligation.
#[test]
fn activity_cancellation_task_round_trips() {
    let task = ActivityTask {
        task_token: "dG9rZW4=".to_owned(),
        variant: ActivityTaskVariant::Cancel(ActivityCancel {
            reason: ActivityCancelReason::WorkerShutdown,
            details: None,
        }),
    };

    let encoded = encode_task(&task).expect("task should encode");
    assert_eq!(decode_task(&encoded), Ok(task));
}

/// Core reports the primary cancellation reason and the independent facts in
/// one asynchronous `ActivityTask::Cancel`.  This fixture exercises each fact
/// that can be set by a heartbeat response, then crosses the exact Core -> Rust
/// conversion and the Rust -> OCaml JSON handoff.  Keeping the assertions on
/// the typed value (rather than only matching JSON text) catches field swaps or
/// accidental derivation of one flag from another.
#[test]
fn activity_cancellation_details_survive_core_to_json_round_trip() {
    let cases = [
        (
            core_activity_task::ActivityCancelReason::Cancelled,
            ActivityCancelReason::Cancelled,
            core_activity_task::ActivityCancellationDetails {
                is_not_found: false,
                is_cancelled: true,
                is_paused: false,
                is_timed_out: false,
                is_worker_shutdown: false,
                is_reset: false,
            },
        ),
        (
            core_activity_task::ActivityCancelReason::Paused,
            ActivityCancelReason::Paused,
            core_activity_task::ActivityCancellationDetails {
                is_not_found: false,
                is_cancelled: false,
                is_paused: true,
                is_timed_out: false,
                is_worker_shutdown: false,
                is_reset: false,
            },
        ),
        (
            core_activity_task::ActivityCancelReason::Reset,
            ActivityCancelReason::Reset,
            core_activity_task::ActivityCancellationDetails {
                is_not_found: false,
                is_cancelled: false,
                is_paused: false,
                is_timed_out: false,
                is_worker_shutdown: false,
                is_reset: true,
            },
        ),
        (
            core_activity_task::ActivityCancelReason::Cancelled,
            ActivityCancelReason::Cancelled,
            core_activity_task::ActivityCancellationDetails {
                is_not_found: false,
                is_cancelled: true,
                is_paused: true,
                is_timed_out: false,
                is_worker_shutdown: false,
                is_reset: true,
            },
        ),
    ];

    for (reason, expected_reason, details) in cases {
        let core = core_activity_task::ActivityTask {
            task_token: vec![0, 1, 2, 254, 255],
            variant: Some(core_activity_task::activity_task::Variant::Cancel(
                core_activity_task::Cancel {
                    reason: reason as i32,
                    details: Some(details),
                },
            )),
        };
        let expected_details =
            ocaml_temporal_core_bridge::activity_protocol::ActivityCancellationDetails {
                is_not_found: details.is_not_found,
                is_cancelled: details.is_cancelled,
                is_paused: details.is_paused,
                is_timed_out: details.is_timed_out,
                is_worker_shutdown: details.is_worker_shutdown,
                is_reset: details.is_reset,
            };

        let semantic = task_from_core(&core).expect("Core cancellation should convert");
        let ActivityTaskVariant::Cancel(cancel) = &semantic.variant else {
            panic!("Core cancellation should remain a Cancel task");
        };
        assert_eq!(cancel.reason, expected_reason);
        assert_eq!(cancel.details, Some(expected_details));

        let encoded = encode_task(&semantic).expect("cancellation task should encode");
        assert_eq!(decode_task(&encoded), Ok(semantic));
    }
}

/// A fully populated start task can cross both directions without changing
/// the typed value, including exact retry-policy bit strings and payloads.
#[test]
fn activity_start_task_round_trips() {
    let task = decode_task(VALID_START_TASK).expect("valid start task should decode");
    let encoded = encode_task(&task).expect("valid start task should encode");
    assert_eq!(decode_task(&encoded), Ok(task));
}

/// Rust rejects the same nested values that the OCaml adapter rejects before
/// either side can invoke an activity with an unsafe execution context.
#[test]
fn activity_start_rejects_nested_semantic_mismatches() {
    let oversized_fairness_key = format!("\"fairness_key\":\"{}\"", "x".repeat(65));
    let oversized_standalone_run_id = format!("\"standalone_run_id\":\"{}\"", "x".repeat(65_537));
    let oversized_non_retryable_type = format!("\"{}\"", "x".repeat(65_537));
    let malformed = [
        (
            "negative retry initial interval",
            replace_once(
                VALID_START_TASK,
                "\"initial_interval\":{\"seconds\":1",
                "\"initial_interval\":{\"seconds\":-1",
            ),
        ),
        (
            "retry maximum interval nanoseconds",
            replace_once(
                VALID_START_TASK,
                "\"maximum_interval\":{\"seconds\":60,\"nanoseconds\":0}",
                "\"maximum_interval\":{\"seconds\":60,\"nanoseconds\":1000000000}",
            ),
        ),
        (
            "noncanonical retry coefficient",
            replace_once(
                VALID_START_TASK,
                "\"backoff_coefficient_bits\":\"4611686018427387904\"",
                "\"backoff_coefficient_bits\":\"01\"",
            ),
        ),
        (
            "retry coefficient outside u64",
            replace_once(
                VALID_START_TASK,
                "\"backoff_coefficient_bits\":\"4611686018427387904\"",
                "\"backoff_coefficient_bits\":\"18446744073709551616\"",
            ),
        ),
        (
            "empty header key",
            replace_once(VALID_START_TASK, "\"trace\":", "\"\":"),
        ),
        (
            "fairness key exceeds Core limit",
            replace_once(
                VALID_START_TASK,
                "\"fairness_key\":\"tenant\"",
                &oversized_fairness_key,
            ),
        ),
        (
            "standalone run ID exceeds text limit",
            replace_once(
                VALID_START_TASK,
                "\"standalone_run_id\":\"\"",
                &oversized_standalone_run_id,
            ),
        ),
        (
            "non-retryable error type exceeds text limit",
            replace_once(
                VALID_START_TASK,
                "\"InvalidInput\"",
                &oversized_non_retryable_type,
            ),
        ),
        (
            "maximum attempts outside signed i32",
            replace_once(
                VALID_START_TASK,
                "\"maximum_attempts\":3",
                "\"maximum_attempts\":2147483648",
            ),
        ),
    ];

    for (name, json) in malformed {
        assert!(decode_task(&json).is_err(), "{name} should be rejected");
    }
}

/// The signed int32 retry-attempt domain is bilateral: both endpoints remain
/// accepted even though a future semantic policy may narrow their meaning.
#[test]
fn retry_maximum_attempts_accepts_signed_i32_boundaries() {
    for value in ["-2147483648", "2147483647"] {
        let json = replace_once(
            VALID_START_TASK,
            "\"maximum_attempts\":3",
            &format!("\"maximum_attempts\":{value}"),
        );
        decode_task(&json).expect("signed i32 boundary should decode");
    }
}

File: lib/protocol/activity_protocol.ml
module Control = Control_protocol
module Workflow = Workflow_protocol
module Shared = Workflow.Internal

type payload = Workflow.payload = { metadata : (string * bytes) list; data : bytes }
type timestamp = Workflow.timestamp = { seconds : int64; nanoseconds : int }
type duration = Workflow.duration = { seconds : int64; nanoseconds : int }
type workflow_execution = Workflow.workflow_execution = { workflow_id : string; run_id : string }
type workflow_priority = Workflow.workflow_priority = { priority_key : int; fairness_key : string; fairness_weight_bits : int64 }

type retry_state = Workflow.retry_state =
  | Unspecified
  | In_progress
  | Non_retryable_failure
  | Timeout
  | Maximum_attempts_reached
  | Retry_policy_not_set
  | Internal_server_error
  | Cancel_requested

type failure_info = Workflow.failure_info =
  | Application of { type_name : string; non_retryable : bool; details : payload list }
  | Canceled of { details : payload list; identity : string }
  | Activity of {
      scheduled_event_id : int64;
      started_event_id : int64;
      identity : string;
      activity_type : string;
      activity_id : string;
      retry_state : retry_state;
    }
  | Child_workflow of {
      namespace : string;
      workflow_id : string;
      run_id : string;
      workflow_type : string;
      initiated_event_id : int64;
      started_event_id : int64;
      retry_state : retry_state;
    }

type failure = Workflow.failure = {
  message : string;
  source : string;
  stack_trace : string;
  encoded_attributes : payload option;
  cause : failure option;
  info : failure_info;
}

type retry_policy = {
  initial_interval : duration option;
  backoff_coefficient_bits : string;
  maximum_interval : duration option;
  maximum_attempts : int;
  non_retryable_error_types : string list;
}

type activity_start = {
  workflow_namespace : string;
  workflow_type : string;
  workflow_execution : workflow_execution;
  activity_id : string;
  activity_type : string;
  header_fields : (string * payload) list;
  input : payload list;
  heartbeat_details : payload list;
  scheduled_time : timestamp option;
  current_attempt_scheduled_time : timestamp option;
  started_time : timestamp option;
  attempt : int64;
  schedule_to_close_timeout : duration option;
  start_to_close_timeout : duration option;
  heartbeat_timeout : duration option;
  retry_policy : retry_policy option;
  priority : workflow_priority option;
  standalone_run_id : string;
}

type cancel_reason =
  | Cancellation_not_found
  | Cancellation_requested
  | Cancellation_timed_out
  | Cancellation_worker_shutdown
  | Cancellation_paused
  | Cancellation_reset

type cancellation_details = {
  is_not_found : bool;
  is_cancelled : bool;
  is_paused : bool;
  is_timed_out : bool;
  is_worker_shutdown : bool;
  is_reset : bool;
}

type activity_cancel = { reason : cancel_reason; details : cancellation_details option }
type task_variant = Start of activity_start | Cancel of activity_cancel
type task = { task_token : bytes; variant : task_variant }

type completion_result =
  | Completed of payload option
  | Failed of failure
  | Cancelled of failure
  | Will_complete_async

type completion = { task_token : bytes; result : completion_result }
type error = Workflow.error
type error_view = Workflow.error_view = { code : string; path : string; message : string }

(** Copies a semantic error without exposing its private representation. *)
let error_view = Workflow.error_view

(** Sequences strict structural and semantic validation without exceptions. *)
let ( let* ) = Result.bind

(** Reads a required object member after its enclosing object has been closed. *)
let required path name entries = Shared.field path name entries

(** Decodes one canonical padded-base64 task token into opaque bytes. *)
let task_token path json =
  match json with
  | `String encoded ->
      (* Tokens deliberately bypass the ordinary 64 KiB text limit. The shared
         binary codec enforces canonical base64 and the 128 MiB byte ceiling. *)
      let wrapper =
        `Assoc [ ("encoding", `String "base64"); ("data", `String encoded) ]
      in
      let* decoded = Shared.bytes_wrapper path wrapper in
      if Bytes.length decoded = 0 then
        Error (Shared.invalid path "activity task token must not be empty")
      else Ok decoded
  | _ -> Error (Shared.invalid path "expected JSON string")

(** Encodes opaque token bytes as the protocol's unwrapped canonical base64
    string. The shared binary codec remains authoritative for size limits. *)
let task_token_json path bytes =
  if Bytes.length bytes = 0 then
    Error (Shared.invalid path "activity task token must not be empty")
  else
    let* wrapper = Shared.bytes_wrapper_json bytes in
    match wrapper with
    | `Assoc entries -> required path "data" entries
    | _ -> Error (Shared.invalid path "canonical task token encoding failed")

(** Decodes a payload map, validates header keys as Temporal identifiers, and
    normalizes association-list order for deterministic typed values. *)
let payload_map path = function
  | `Assoc entries ->
      let rec loop decoded = function
        | [] ->
            Ok
              (List.sort
                 (fun (left, _) (right, _) -> String.compare left right)
                 decoded)
        | (key, value) :: rest ->
            let* _ = Shared.identifier (path ^ ".<key>") (`String key) in
            let* value = Shared.payload (path ^ "." ^ key) value in
            loop ((key, value) :: decoded) rest
      in
      loop [] entries
  | _ -> Error (Shared.invalid path "expected JSON object")

(** Encodes a payload map while rejecting keys that are invalid or repeated in
    OCaml's association-list representation. *)
let payload_map_json path values =
  let rec loop encoded = function
    | [] -> Ok (`Assoc encoded)
    | (key, value) :: rest ->
        let* _ = Shared.identifier (path ^ ".<key>") (`String key) in
        if List.exists (fun (existing, _) -> String.equal existing key) encoded then
          Error (Shared.invalid path "duplicate payload-map key")
        else
          let* value = Shared.payload_json value in
          loop ((key, value) :: encoded) rest
  in
  loop [] values

(** Decodes a list of ordinary bounded strings. *)
let strings path json = Shared.list path Shared.string json

(** Encodes an optional timestamp as an explicitly present null. *)
let optional_timestamp (value : timestamp option) =
  match value with
  | None -> `Null
  | Some value -> Shared.time_json value.seconds value.nanoseconds

(** Encodes an optional duration as an explicitly present null. *)
let optional_duration (value : duration option) =
  match value with
  | None -> `Null
  | Some value -> Shared.time_json value.seconds value.nanoseconds

(** Decodes the retry policy exactly as represented by Rust, including an
    unsigned decimal string for the raw [f64] backoff coefficient bits. *)
let retry_policy path json =
  let fields =
    [
      "initial_interval";
      "backoff_coefficient_bits";
      "maximum_interval";
      "maximum_attempts";
      "non_retryable_error_types";
    ]
  in
  let* entries = Shared.exact_object path fields json in
  let* initial_json = required path "initial_interval" entries in
  let* initial_interval = Shared.nullable (path ^ ".initial_interval") Shared.duration initial_json in
  let* backoff_json = required path "backoff_coefficient_bits" entries in
  let* backoff_coefficient_bits =
    Shared.uint64_decimal (path ^ ".backoff_coefficient_bits") backoff_json
  in
  let* maximum_json = required path "maximum_interval" entries in
  let* maximum_interval = Shared.nullable (path ^ ".maximum_interval") Shared.duration maximum_json in
  let* attempts_json = required path "maximum_attempts" entries in
  let* maximum_attempts = Shared.int32 (path ^ ".maximum_attempts") attempts_json in
  let* errors_json = required path "non_retryable_error_types" entries in
  let* non_retryable_error_types = strings (path ^ ".non_retryable_error_types") errors_json in
  Ok
    {
      initial_interval;
      backoff_coefficient_bits;
      maximum_interval;
      maximum_attempts;
      non_retryable_error_types;
    }

(** Encodes one retry policy; the complete document is reparsed so sender-side
    values receive exactly the same numeric validation as received JSON. *)
let retry_policy_json value =
  Ok
    (`Assoc
      [
        ("initial_interval", optional_duration value.initial_interval);
        ("backoff_coefficient_bits", `String value.backoff_coefficient_bits);
        ("maximum_interval", optional_duration value.maximum_interval);
        ("maximum_attempts", `Int value.maximum_attempts);
        ( "non_retryable_error_types",
          `List (List.map (fun value -> `String value) value.non_retryable_error_types) );
      ])

(** Renders one workflow execution identity using the shared record model. *)
let workflow_execution_json value =
  `Assoc
    [
      ("workflow_id", `String value.workflow_id);
      ("run_id", `String value.run_id);
    ]

(** Renders one workflow priority without converting its weight bits to float. *)
let workflow_priority_json value =
  `Assoc
    [
      ("priority_key", `Int value.priority_key);
      ("fairness_key", `String value.fairness_key);
      ("fairness_weight_bits", `Intlit (Int64.to_string value.fairness_weight_bits));
    ]

(** Decodes complete execution context for one start-task attempt. *)
let activity_start path json =
  let fields =
    [
      "kind";
      "workflow_namespace";
      "workflow_type";
      "workflow_execution";
      "activity_id";
      "activity_type";
      "header_fields";
      "input";
      "heartbeat_details";
      "scheduled_time";
      "current_attempt_scheduled_time";
      "started_time";
      "attempt";
      "schedule_to_close_timeout";
      "start_to_close_timeout";
      "heartbeat_timeout";
      "retry_policy";
      "priority";
      "standalone_run_id";
    ]
  in
  let* entries = Shared.exact_object path fields json in
  let get name = required path name entries in
  let* workflow_namespace_json = get "workflow_namespace" in
  let* workflow_namespace = Shared.identifier (path ^ ".workflow_namespace") workflow_namespace_json in
  let* workflow_type_json = get "workflow_type" in
  let* workflow_type = Shared.identifier (path ^ ".workflow_type") workflow_type_json in
  let* execution_json = get "workflow_execution" in
  let* workflow_execution = Shared.workflow_execution (path ^ ".workflow_execution") execution_json in
  let* activity_id_json = get "activity_id" in
  let* activity_id = Shared.identifier (path ^ ".activity_id") activity_id_json in
  let* activity_type_json = get "activity_type" in
  let* activity_type = Shared.identifier (path ^ ".activity_type") activity_type_json in
  let* headers_json = get "header_fields" in
  let* header_fields = payload_map (path ^ ".header_fields") headers_json in
  let* input_json = get "input" in
  let* input = Shared.list (path ^ ".input") Shared.payload input_json in
  let* heartbeat_json = get "heartbeat_details" in
  let* heartbeat_details = Shared.list (path ^ ".heartbeat_details") Shared.payload heartbeat_json in
  let nullable_time name decoder =
    let* json = get name in
    Shared.nullable (path ^ "." ^ name) decoder json
  in
  let* scheduled_time = nullable_time "scheduled_time" Shared.timestamp in
  let* current_attempt_scheduled_time = nullable_time "current_attempt_scheduled_time" Shared.timestamp in
  let* started_time = nullable_time "started_time" Shared.timestamp in
  let* attempt_json = get "attempt" in
  let* attempt = Shared.uint32 (path ^ ".attempt") attempt_json in
  let* schedule_to_close_timeout = nullable_time "schedule_to_close_timeout" Shared.duration in
  let* start_to_close_timeout = nullable_time "start_to_close_timeout" Shared.duration in
  let* heartbeat_timeout = nullable_time "heartbeat_timeout" Shared.duration in
  let* retry_json = get "retry_policy" in
  let* retry_policy = Shared.nullable (path ^ ".retry_policy") retry_policy retry_json in
  let* priority_json = get "priority" in
  let* priority = Shared.nullable (path ^ ".priority") Shared.workflow_priority priority_json in
  let* standalone_json = get "standalone_run_id" in
  let* standalone_run_id = Shared.string (path ^ ".standalone_run_id") standalone_json in
  Ok
    {
      workflow_namespace;
      workflow_type;
      workflow_execution;
      activity_id;
      activity_type;
      header_fields;
      input;
      heartbeat_details;
      scheduled_time;
      current_attempt_scheduled_time;
      started_time;
      attempt;
      schedule_to_close_timeout;
      start_to_close_timeout;
      heartbeat_timeout;
      retry_policy;
      priority;
      standalone_run_id;
    }

(** Encodes a payload list while preserving argument order. *)
let payloads_json values =
  let rec loop encoded = function
    | [] -> Ok (`List (List.rev encoded))
    | value :: rest ->
        let* value = Shared.payload_json value in
        loop (value :: encoded) rest
  in
  loop [] values

(** Encodes complete start-task context for strict sender-side reparsing. *)
let activity_start_json value =
  let* header_fields = payload_map_json "$.variant.header_fields" value.header_fields in
  let* input = payloads_json value.input in
  let* heartbeat_details = payloads_json value.heartbeat_details in
  let* retry_policy =
    match value.retry_policy with
    | None -> Ok `Null
    | Some value -> retry_policy_json value
  in
  let priority =
    match value.priority with None -> `Null | Some value -> workflow_priority_json value
  in
  Ok
    (`Assoc
      [
        ("kind", `String "start");
        ("workflow_namespace", `String value.workflow_namespace);
        ("workflow_type", `String value.workflow_type);
        ("workflow_execution", workflow_execution_json value.workflow_execution);
        ("activity_id", `String value.activity_id);
        ("activity_type", `String value.activity_type);
        ("header_fields", header_fields);
        ("input", input);
        ("heartbeat_details", heartbeat_details);
        ("scheduled_time", optional_timestamp value.scheduled_time);
        ("current_attempt_scheduled_time", optional_timestamp value.current_attempt_scheduled_time);
        ("started_time", optional_timestamp value.started_time);
        ("attempt", `Intlit (Int64.to_string value.attempt));
        ("schedule_to_close_timeout", optional_duration value.schedule_to_close_timeout);
        ("start_to_close_timeout", optional_duration value.start_to_close_timeout);
        ("heartbeat_timeout", optional_duration value.heartbeat_timeout);
        ("retry_policy", retry_policy);
        ("priority", priority);
        ("standalone_run_id", `String value.standalone_run_id);
      ])

(** Decodes the stable cancellation reason spelling. *)
let cancel_reason path = function
  | "not_found" -> Ok Cancellation_not_found
  | "cancelled" -> Ok Cancellation_requested
  | "timed_out" -> Ok Cancellation_timed_out
  | "worker_shutdown" -> Ok Cancellation_worker_shutdown
  | "paused" -> Ok Cancellation_paused
  | "reset" -> Ok Cancellation_reset
  | _ -> Error (Shared.invalid path "unknown activity cancellation reason")

(** Renders the stable cancellation reason spelling. *)
let cancel_reason_string = function
  | Cancellation_not_found -> "not_found"
  | Cancellation_requested -> "cancelled"
  | Cancellation_timed_out -> "timed_out"
  | Cancellation_worker_shutdown -> "worker_shutdown"
  | Cancellation_paused -> "paused"
  | Cancellation_reset -> "reset"

(** Decodes the independent cancellation facts as one closed object. *)
let cancellation_details path json =
  let fields =
    [
      "is_not_found";
      "is_cancelled";
      "is_paused";
      "is_timed_out";
      "is_worker_shutdown";
      "is_reset";
    ]
  in
  let* entries = Shared.exact_object path fields json in
  let read name =
    let* json = required path name entries in
    Shared.bool (path ^ "." ^ name) json
  in
  let* is_not_found = read "is_not_found" in
  let* is_cancelled = read "is_cancelled" in
  let* is_paused = read "is_paused" in
  let* is_timed_out = read "is_timed_out" in
  let* is_worker_shutdown = read "is_worker_shutdown" in
  let* is_reset = read "is_reset" in
  Ok
    {
      is_not_found;
      is_cancelled;
      is_paused;
      is_timed_out;
      is_worker_shutdown;
      is_reset;
    }

(** Encodes independent cancellation facts without deriving one flag from
    another; Core treats each field as separate information. *)
let cancellation_details_json value =
  `Assoc
    [
      ("is_not_found", `Bool value.is_not_found);
      ("is_cancelled", `Bool value.is_cancelled);
      ("is_paused", `Bool value.is_paused);
      ("is_timed_out", `Bool value.is_timed_out);
      ("is_worker_shutdown", `Bool value.is_worker_shutdown);
      ("is_reset", `Bool value.is_reset);
    ]

(** Decodes either a start or cancellation task from the closed tagged union. *)
let task_variant path json =
  let* entries =
    match json with
    | `Assoc entries -> Ok entries
    | _ -> Error (Shared.invalid path "expected JSON object")
  in
  let* kind_json = required path "kind" entries in
  let* kind = Shared.string (path ^ ".kind") kind_json in
  match kind with
  | "start" ->
      let* value = activity_start path json in
      Ok (Start value)
  | "cancel" ->
      let* entries = Shared.exact_object path [ "kind"; "reason"; "details" ] json in
      let* reason_json = required path "reason" entries in
      let* reason_name = Shared.string (path ^ ".reason") reason_json in
      let* reason = cancel_reason (path ^ ".reason") reason_name in
      let* details_json = required path "details" entries in
      let* details = Shared.nullable (path ^ ".details") cancellation_details details_json in
      Ok (Cancel { reason; details })
  | _ -> Error (Shared.invalid (path ^ ".kind") "unknown activity task kind")

(** Encodes either task variant using its exact closed member set. *)
let task_variant_json = function
  | Start value -> activity_start_json value
  | Cancel value ->
      let details =
        match value.details with
        | None -> `Null
        | Some value -> cancellation_details_json value
      in
      Ok
        (`Assoc
          [
            ("kind", `String "cancel");
            ("reason", `String (cancel_reason_string value.reason));
            ("details", details);
          ])

(** Converts one strict task object into typed activity execution data. *)
let task_from_json json =
  let* entries = Shared.exact_object "$" [ "task_token"; "variant" ] json in
  let* token_json = required "$" "task_token" entries in
  let* task_token = task_token "$.task_token" token_json in
  let* variant_json = required "$" "variant" entries in
  let* variant = task_variant "$.variant" variant_json in
  Ok { task_token; variant }

(** Strictly decodes a task through the duplicate-aware JSON foundation. *)
let decode_task input =
  match Control.decode_payload_object input with
  | Error error -> Error (Shared.of_control_error "$" error)
  | Ok json -> task_from_json json

(** Encodes and reparses a task so typed outgoing values obey every receiver
    invariant before crossing the native boundary. *)
let encode_task (value : task) =
  let* token = task_token_json "$.task_token" value.task_token in
  let* variant = task_variant_json value.variant in
  let json = `Assoc [ ("task_token", token); ("variant", variant) ] in
  match Control.encode_payload_object json with
  | Error error -> Error (Shared.of_control_error "$" error)
  | Ok output ->
      let* _ = decode_task output in
      Ok output

(** Decodes one closed activity completion result. *)
let completion_result path json =
  let* entries =
    match json with
    | `Assoc entries -> Ok entries
    | _ -> Error (Shared.invalid path "expected JSON object")
  in
  let* kind_json = required path "kind" entries in
  let* kind = Shared.string (path ^ ".kind") kind_json in
  match kind with
  | "completed" ->
      let* entries = Shared.exact_object path [ "kind"; "result" ] json in
      let* result_json = required path "result" entries in
      let* result = Shared.nullable (path ^ ".result") Shared.payload result_json in
      Ok (Completed result)
  | "failed" | "cancelled" ->
      let* entries = Shared.exact_object path [ "kind"; "failure" ] json in
      let* failure_json = required path "failure" entries in
      let* failure = Shared.failure (path ^ ".failure") failure_json in
      if String.equal kind "failed" then Ok (Failed failure)
      else Ok (Cancelled failure)
  | "will_complete_async" ->
      let* _ = Shared.exact_object path [ "kind" ] json in
      Ok Will_complete_async
  | _ -> Error (Shared.invalid (path ^ ".kind") "unknown activity completion kind")

(** Encodes one terminal completion result through the shared payload or
    failure codec as appropriate. *)
let completion_result_json = function
  | Completed result ->
      let* result =
        match result with None -> Ok `Null | Some value -> Shared.payload_json value
      in
      Ok (`Assoc [ ("kind", `String "completed"); ("result", result) ])
  | Failed failure ->
      let* failure = Shared.failure_json failure in
      Ok (`Assoc [ ("kind", `String "failed"); ("failure", failure) ])
  | Cancelled failure ->
      let* failure = Shared.failure_json failure in
      Ok (`Assoc [ ("kind", `String "cancelled"); ("failure", failure) ])
  | Will_complete_async -> Ok (`Assoc [ ("kind", `String "will_complete_async") ])

(** Converts one strict completion object into a typed terminal response. *)
let completion_from_json json =
  let* entries = Shared.exact_object "$" [ "task_token"; "result" ] json in
  let* token_json = required "$" "task_token" entries in
  let* task_token = task_token "$.task_token" token_json in
  let* result_json = required "$" "result" entries in
  let* result = completion_result "$.result" result_json in
  Ok { task_token; result }

(** Strictly decodes a completion through the duplicate-aware JSON foundation. *)
let decode_completion input =
  match Control.decode_payload_object input with
  | Error error -> Error (Shared.of_control_error "$" error)
  | Ok json -> completion_from_json json

(** Encodes and reparses a completion before it can be submitted to Core. *)
let encode_completion (value : completion) =
  let* token = task_token_json "$.task_token" value.task_token in
  let* result = completion_result_json value.result in
  let json = `Assoc [ ("task_token", token); ("result", result) ] in
  match Control.encode_payload_object json with
  | Error error -> Error (Shared.of_control_error "$" error)
  | Ok output ->
      let* _ = decode_completion output in
      Ok output
